api/test/trace/trace_state_test.cc

```diff
@@ -60,81 +61,129 @@ TEST(EntryTest, SetValue)
-  opentelemetry::nostd::string_view null_val;
-  // Since string_view data is null by default, this should be a no-op
-  s.Set(bad_key, null_val);
+  trace_state_header = "k1=v1,k2=v2,k3=v3";
+  EXPECT_EQ(create_ts_return_header(trace_state_header), trace_state_header);
 
-  opentelemetry::nostd::span<TraceState::Entry> entries = s.Entries();
-  EXPECT_EQ(entries.size(), 1);
-  EXPECT_EQ(entries[0].GetKey(), key);
-  EXPECT_EQ(entries[0].GetValue(), val);
+  trace_state_header = "k1=v1,k2=v2,k3=v3";
+  EXPECT_EQ(create_ts_return_header(trace_state_header), trace_state_header);
+
+  trace_state_header         = "k1=v1,k2=v2,,";
+  auto expected_state_header = "k1=v1,k2=v2";
+  EXPECT_EQ(create_ts_return_header(trace_state_header), expected_state_header);
+
+  trace_state_header = "k1=v1,k2=v2,sd";                       // invalid list member
+  EXPECT_EQ(create_ts_return_header(trace_state_header), "");  // empty header
+
+  trace_state_header = "1a-2f@foo=bar1,a*/foo-_/bar=bar4";  // valid list
+  EXPECT_EQ(create_ts_return_header(trace_state_header), trace_state_header);
+
+  trace_state_header = "1a-2f@foo=bar1,*/foo-_/bar=bar4";  // invalid list member, key `*/foo-_/bar`
+                                                           // must begin with [a-z,0-9]
+  EXPECT_EQ(create_ts_return_header(trace_state_header), "");
+
+  trace_state_header = header_with_max_members();
+  EXPECT_EQ(create_ts_return_header(trace_state_header), trace_state_header);
 }
 
-TEST(TraceStateTest, Get)
+TEST(TraceStateTest, TraceStateGet)
 {
-  TraceState s;
-  const int kNumPairs                                 = 3;
-  opentelemetry::nostd::string_view keys[kNumPairs]   = {"test_key_1", "test_key_2", "test_key_3"};
-  opentelemetry::nostd::string_view values[kNumPairs] = {"test_val_1", "test_val_2", "test_val_3"};
 
-  for (int i = 0; i < kNumPairs; i++)
-  {
-    s.Set(keys[i], values[i]);
-  }
+  std::string trace_state_header = header_with_max_members();
+  TraceState ts                  = TraceState::FromHeader(trace_state_header);
 
-  opentelemetry::nostd::string_view return_val = "";
+  EXPECT_EQ(ts.Get("key0"), "value0");
+  EXPECT_EQ(ts.Get("key16"), "value16");
+  EXPECT_EQ(ts.Get("key31"), "value31");
+  EXPECT_EQ(ts.Get("key32"), "");  // key not found
+}
 
-  for (int i = 0; i < kNumPairs; i++)
-  {
-    EXPECT_TRUE(s.Get(keys[i], return_val));
-    EXPECT_EQ(return_val, values[i]);
-    return_val = "";
-  }
+TEST(TraceStateTest, TraceStateSet)
+{
+  std::string trace_state_header = "k1=v1,k2=v2";
+  auto ts1                       = TraceState::FromHeader(trace_state_header);
+  auto ts1_new                   = ts1.Set("k3", "v3");
+  EXPECT_EQ(ts1_new.ToHeader(), "k3=v3,k1=v1,k2=v2");
+
+  trace_state_header = header_with_max_members();
+  auto ts2           = TraceState::FromHeader(trace_state_header);
+  auto ts2_new       = ts2.Set("n_k1", "n_v1");  // adding to max list, should fail and return empty
+  EXPECT_EQ(ts2_new.ToHeader(), "");
```

Comment:
What consequence would this behavior have for the user?

Let's say one receives a `tracestate` header with 32 entries, then adds one (this returns an empty `tracestate`), and then passes on the header. If the user doesn't do explicit checks, an empty header will be passed on. I don't think that's ideal.

What do you think about `Set` just returning the unaltered `tracestate` when the limit is exceeded?

Reply:
Yes, I didn't find anything in specs on how to address this, and different languages implementation are behaving differently ( python returns empty, JS doesn't do any check on limit, java returns reference to itself ). Said that, it make sense to return unaltered `tracestate` instead of empty. I will do the change. Thanks for the comment.

Reply:
This is fixed as discussed.